### DISCUSSION TITLE
Watch an image file in wezterm?

### DISCUSSION THREAD

**Discussion by chtenb** [wezterm/wezterm#5562] [Q&A] [answered]
### What Operating System(s) are you running on?

Windows

### WezTerm version

latest

### Ask your question!

Would it be currently be possible to utilize `wezterm imgcat` in such a way that you could watch an image file? Kind of like `tail -f` but for images. The use case I'm looking at is for diagram writing and rendering. I would love to be able to write the diagram text in one pane, while having a watch command on the other pane, refreshing the rendered result as I make changes.

**Comment by wez** [comment, marked as answer]
maybe `watch wezterm imgcat file.png`

https://www.man7.org/linux/man-pages/man1/watch.1.html

Since you're on windows and may not have `watch`, what that does is essentially a loop that sleeps for a couple of seconds then runs the command.

If you want something more sophisticated, that does something like using OS-level file monitoring, then maybe something like https://github.com/thekid/inotify-win will work for you

**Reply by chtenb** [reply]
Hm, I do have a watch command available through nushell. However, invoking it seems to only dump a black rectangle followed by a lot of ascii characters on the screen.
```
watch -d 2000 services.png { || wezterm imgcat services.png }
```

`wezterm imgcat services.png` on its own works fine.

**Reply by nekowinston** [reply]
@chtenb the closure tries to cast to a string with having the external command (`wezterm imgcat`) as the last statement. 
Try to handle it like so:
```nushell
watch -d 2000 ./foo.jpg { |_, path| wezterm imgcat $path; return }
```

**Reply by chtenb** [reply]
Thanks so much to the both of you. Works like a charm!

**Reply by chtenb** [reply]
Just for reference, I've defined a nu command for this which calls plantuml on file change and displays it to the screen in some given resolution. It would be even cooler if I could parameterize the 1600x1200 automatically based on terminal resolutions, but I haven't found a clean way to do that yet.
```nu
def pumlwatch [file] {
  watch -d 100 $file { |_, path| cat $path | plantuml -p -tsvg | tee {save -f $"($path).svg"}
    | magick.exe svg:- -resize 1600x1200 png:- | wezterm imgcat; return }
}
```

**Reply by wez** [reply]
FWIW, `wezterm cli list --format=json` will show you the dimensions of each pane:

```json
[
  {
    "window_id": 3,
    "tab_id": 16,
    "pane_id": 34,
    "workspace": "default",
    "size": {
      "rows": 81,
      "cols": 227,
      "pixel_width": 2270,
      "pixel_height": 1701,
      "dpi": 120
    },
    "title": "wez@foo:~/wez-personal/wezterm",
    "cwd": "file://foo/home/wez/wez-personal/wezterm",
    "cursor_x": 2,
    "cursor_y": 80,
    "cursor_shape": "Default",
    "cursor_visibility": "Visible",
    "left_col": 0,
    "top_row": 0,
    "tab_title": "",
    "window_title": "wezterm cli list --format=json -- wez@foo:~/wez-personal/integrator",
    "is_active": true,
    "is_zoomed": false,
    "tty_name": "/dev/pts/5"
  }
]
```

You can look up the `pane_id` of the current pane by looking at `$WEZTERM_PANE`.

Caveat emptor: on Windows, the kernel doesn't have a native understanding of the pixel dimensions so there may be some cases where `size.pixel_width` and `size.pixel_height` are inaccurate.

**Reply by chtenb** [reply]
Nice! That means I can do

```nu
def pumlwatch [file] {
  watch $file { |op, path| if ($op == "Write") { 
      let size = wezterm cli list --format=json | from json 
        | where pane_id == ($env.WEZTERM_PANE | into int) | get 0.size
      cat $path | plantuml -p -tsvg | tee {save -f $"($path).svg"}
        | magick.exe svg:- -resize $"($size.pixel_width)x($size.pixel_height)" png:- | wezterm imgcat
      return
    }
  }
}

```
to make the image automatically scale up to the pane dimensions (preserving aspect ratio)
